pkg/importer/imageio-datasource.go

```diff
@@ -243,30 +310,183 @@ func createImageioReader(ctx context.Context, ep string, accessKey string, secKe
+			ovirtsdk4.IMAGETRANSFERPHASE_INITIALIZING:       cancelTransfer,
+			ovirtsdk4.IMAGETRANSFERPHASE_PAUSED_SYSTEM:      cancelTransfer,
+			ovirtsdk4.IMAGETRANSFERPHASE_PAUSED_USER:        cancelTransfer,
+			ovirtsdk4.IMAGETRANSFERPHASE_RESUMING:           cancelTransfer,
+			ovirtsdk4.IMAGETRANSFERPHASE_TRANSFERRING:       finalizeTransfer,
+			ovirtsdk4.IMAGETRANSFERPHASE_UNKNOWN:            cancelTransfer,
+			// Observed from RHV, but not yet listed in Go API:
+			"cancelled_system":   nil,
+			"cancelled_user":     nil,
+			"finalizing_cleanup": nil,
+		}
+		action, available := phaseActions[transferPhase]
+		if !available {
+			klog.Warningf("Unknown transfer phase '%s', %d retries remaining.", transferPhase, retries)
+			cancelTransfer()
+			time.Sleep(delay)
+			continue
+		}
+		if action != nil {
+			err = action()
+			if err != nil {
+				klog.Warningf("Failed to run transfer cleanup command, %d retries remaining.", retries)
+				time.Sleep(delay)
+				continue
+			}
+		} else {
+			klog.Info("No cleanup action required for this image transfer phase, done.")
+			return nil
 		}
 	}
-	return err
+
+	return errors.Wrapf(err, "retry limit exceeded for transfer ticket cleanup, disk may remain locked until inactivity timeout")
 }
 
-func getTransfer(conn ConnectionInterface, diskID string) (*ovirtsdk4.ImageTransfer, uint64, error) {
+// getDisk finds the disk with the given ID
+func getDisk(conn ConnectionInterface, diskID string) (*ovirtsdk4.Disk, error) {
 	disksService := conn.SystemService().DisksService()
 	diskService := disksService.DiskService(diskID)
 	diskRequest := diskService.Get()
 	diskResponse, err := diskRequest.Send()
 	if err != nil {
-		return nil, uint64(0), errors.Wrap(err, "Error fetching disk")
+		return nil, errors.Wrapf(err, "error fetching disk %s", diskID)
 	}
 	disk, success := diskResponse.Disk()
 	if !success {
-		return nil, uint64(0), errors.New("Error disk not found")
+		return nil, errors.Errorf("disk %s not found", diskID)
+	}
+	id, success := disk.Id()
+	if !success {
+		klog.Warningf("Unable to get ID from disk object, setting it from given %s.", diskID)
+		disk.SetId(diskID)
+	} else if id != diskID {
+		klog.Warningf("Retrieved disk ID %s does not match expected disk ID %s!", id, diskID)
+	}
+	return disk, nil
+}
+
+// getSnapshot finds the snapshot with the given ID on the given disk.
+func getSnapshot(conn ConnectionInterface, disk *ovirtsdk4.Disk, snapshotID string) (*ovirtsdk4.DiskSnapshot, error) {
+	diskID, _ := disk.Id()
+	storageDomains, success := disk.StorageDomains()
+	if !success {
+		return nil, errors.Errorf("no storage domains listed for disk %s", diskID)
+	}
+
+	for index, storageDomain := range storageDomains.Slice() {
+		storageDomainName, success := storageDomain.Id()
+		if !success {
+			klog.Warningf("Skipping storage domain #%d listed on disk %s", index, diskID)
+			continue
+		}
+
+		storageDomainsService := conn.SystemService().StorageDomainsService()
```

Comment:
Can this ever be nil? If so the next line is a potential npe. If not, you can just chain like this:
```go
storageDomainService := conn.SystemService().StorageDomainsService().StorageDomainService(storageDomainName)
```
Same question for storageDomainService, which seems to only be used to find snapshots. And again for diskSnapshotsService, which seems to only be used to list the requests.

Reply:
If `conn` is not nil, then the rest will AFAIK never be nil. These are purely Java-esque constructs which we can't change now without breaking compatibility.

Reply:
Thanks Janos, I chained most of them.

Reply:
That's fine. The calls that need extra checking usually have a second boolean return parameter that tells you if the object was present in the response.